### ISSUE TITLE
Possible regression in alwaysUp alwaysDown

### ISSUE THREAD

**Issue by ssandon** [shred/commons-suncalc#14]
Hi,

Latest correction introduced in 2.5 seems to have modified the alwaysDown or alwaysUp computation (I spotted few changes in the code).

```
// Sydney
lat=-33.744272
lng=151.231291
java.util.GregorianCalendar[time=1562121386435,areFieldsSet=false,areAllFieldsSet=false,lenient=true,zone=libcore.util.ZoneInfo[id="Australia/Sydney",mRawOffset=36000000,mEarliestRawOffset=36000000,mUseDst=true,mDstSavings=3600000,transitions=141],firstDayOfWeek=1,minimalDaysInFirstWeek=1,ERA=1,YEAR=2019,MONTH=6,WEEK_OF_YEAR=27,WEEK_OF_MONTH=1,DAY_OF_MONTH=3,DAY_OF_YEAR=184,DAY_OF_WEEK=4,DAY_OF_WEEK_IN_MONTH=1,AM_PM=0,HOUR=4,HOUR_OF_DAY=4,MINUTE=36,SECOND=26,MILLISECOND=435,ZONE_OFFSET=3600000,DST_OFFSET=3600000])

SunTimes suntimesToday = SunTimes.compute().on(calendar).timezone(tz).midnight().at(latitude, longitude).execute(); // 1

SunTimes suntimesNow = SunTimes.compute().on(calendar).timezone(tz).at(latitude, longitude).execute(); // 2
```
// 1 : is used to compute 'day' lenght for current sun day cycle.

// 2 : only 'usefull' if observer is in the same timezone I guess.

 output is :       

v2.4
suntimesToday=SunTimes[rise=Tue Jul 02 23:01:00 GMT+02:00 2019, set=Wed Jul 03 08:58:00 GMT+02:00 2019, noon=Wed Jul 03 04:02:00 GMT+02:00 2019, nadir=Wed Jul 03 16:40:00 GMT+02:00 2019, alwaysUp=**false**, alwaysDown=**false**]
-> the rise occured yesterday but the current observer date seats in the middle.

suntimesNow=SunTimes[rise=Wed Jul 03 23:00:00 GMT+02:00 2019, set=Wed Jul 03 08:58:00 GMT+02:00 2019, noon=Thu Jul 04 03:59:00 GMT+02:00 2019, nadir=Wed Jul 03 15:59:00 GMT+02:00 2019, alwaysUp=**false**, alwaysDown=**false**]
-> this one is not interesting because it mixes 2 differents cycle.

v2.5
suntimesToday=SunTimes[rise=Tue Jul 02 23:01:00 GMT+02:00 2019, set=Wed Jul 03 08:58:00 GMT+02:00 2019, noon=Wed Jul 03 03:59:00 GMT+02:00 2019, nadir=null, alwaysUp=**true**, alwaysDown=**false**]
2019-07-03 04:59:55.262 21608-21608/org.ssandon.map3d D/Sun: suntimesNow=SunTimes[rise=Wed Jul 03 23:00:00 GMT+02:00 2019, set=Wed Jul 03 08:58:00 GMT+02:00 2019, noon=Thu Jul 04 03:59:00 GMT+02:00 2019, nadir=Wed Jul 03 15:59:00 GMT+02:00 2019, alwaysUp=**false**, alwaysDown=**false**]

My goal is to compute the length of the day (opposed to night time).
I thought alwaysUp or alwaysDown could be usefull for special case above polar circle.
But maybe I am wrong. I could just check if SunTimes.rise or set return something not null ?

Thank you
Stéphane

**Comment by shred** [comment]
Good catch!

Yes, alwaysUp and alwaysDown is for above the polar circle. Otherwise, if there was no sunrise and sunset time on the given day, one wouldn't know if the sun is currently up or down.

I have assumed that the sun is "always up" if there is no solar midnight on that day. Which was a stupid assumption, of course. :man_facepalming:

Because of the fix of issue #13, the calculated nadirs on Jul 03 in Sydney are now slightly before 00:00 and slightly after 24:00, giving no actual solar midnight on that date. As result, the sun was assumed to be "always up" even though there was a sunset time.

There is a fix for this bug in commit e7552dc. The sun is now "always up" if it stays above the twilight angle for 24 hours.

Thank you for taking the time to report this issue.

**Comment by ssandon** [comment]
Thank you for looking at this.
Naturally I would have put something like this

`
// init with default normal case
isAlwaysUp = false;
isAlwaysDown = false;

computation ...

if ((sunrise != null)||(sunset != null)) {
	// 3 cases :
	// sunrise && sunset : normal
	// sunrise only : equinox
	// sunset only : equinox
	
	isAlwaysUp = false;
	isAlwaysDown = false;
} else {
	// We know their isn't sunrise and sunset
	// looking at the current position now tell us the status
	
	if (altitude >= RISE_ALTITUDE) {
		isAlwaysUp = true;
	} else {
		isAlwaysDown = true;
	}
}`

you can get ride of the first test by negating the expression. 

// init with default normal case
isAlwaysUp = false;
isAlwaysDown = false;

computation ...

if (sunrise == null)&&(sunset == null)) {
	// We know their isn't sunrise and sunset
	// looking at the current position now tell us the status
	
	if (altitude >= RISE_ALTITUDE) {
		isAlwaysUp = true;
	} else {
		isAlwaysDown = true;
	}
}`

Regards
Stéphane

**Comment by ssandon** [comment]
If I understand correctly : 
the state (isAlwaysUp == true) and (isAlwaysDown == true) couldn't happens.
So this is why I initialized them to false by default.

**Comment by shred** [comment]
You are right, initializing both variables to `true` is confusing. For the calculation, the oneDay/fullCycle mode has to be taken into account. On fullCycle, alwaysUp and alwaysDown will only reflect the first 24 hours, so I added a check. See commit 8705e38.

I have just released suncalc v2.6 with the fix. Thanks again for finding it! :smile:
